<a href="https://colab.research.google.com/github/erinmcmahon26/NLP-Learning/blob/main/TF_IDF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

How do we decide which words are more important in a single document and accross an entire corpus? This importance will help decide relevant documents in a corpus based on keyword importance within each document. The goal is to find numerical representation of words that capture the importance of the words they represent.

### Bag of Words

In [46]:
sentence = "Mary had a little lamb and the little lamb didn't like the rain."

In [23]:
import nltk
#nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [48]:
import string
import re
from nltk.tokenize import TreebankWordTokenizer

def clean_doc(doc):
# split into tokens by white space
  tokenizer = TreebankWordTokenizer()
  tokens = tokenizer.tokenize(sentence.lower())
# prepare regex for char filtering
  re_punc = re.compile('[%s]' % re.escape(string.punctuation)) # remove punctuation from each word
  tokens = [re_punc.sub('', w) for w in tokens]
# remove remaining tokens that are not alphabetic
  tokens = [word for word in tokens if word.isalpha()]
# filter out stop words
  stop_words = set(stopwords.words('english'))
  tokens = [w for w in tokens if not w in stop_words]
# filter out short tokens
  tokens = [word for word in tokens if len(word) > 1]
  return tokens

In [51]:
tokens = clean_doc(sentence)
tokens

['mary', 'little', 'lamb', 'little', 'lamb', 'nt', 'like', 'rain']

In [52]:
from collections import Counter
bow = Counter(tokens)
bow.most_common(5)

[('little', 2), ('lamb', 2), ('mary', 1), ('nt', 1), ('like', 1)]

The amount of times a word shows up in a document, is called term frequency (TF). Here the term frequency of 'little' is 2. Here we still have 'nt' as a token after using the TreebankWordTokenizer which may or may not be a problem later. We will keep it for now to focus on TF.

In [53]:
tf = bow['little']/len(bow)
round(tf, 4)

0.3333

0.3333 is the # of unique tokens from the original document (sentence). This helps you look at the importance of the word within a specific document versus just the actual word count. 